pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+	if clusterDeployment.Spec.Hardware.AWS == nil {
+		return fmt.Errorf("AWS hardware is not specified in clusterdeployment %s/%s", clusterDeployment.Namespace, clusterDeployment.Name)
+	}
+
+	if clusterVersion.Spec.VMImages.AWSImages == nil {
+		return fmt.Errorf("No AWS images specified in clusterVersion %s/%s", clusterVersion.Namespace, clusterVersion.Name)
+	}
+
+	// Make sure the cluster version supports the region for the clusterDeployment, if not return an error
+	var foundRegion bool
+	for _, regionAMI := range clusterVersion.Spec.VMImages.AWSImages.RegionAMIs {
+		if regionAMI.Region == clusterDeployment.Spec.Hardware.AWS.Region {
+			foundRegion = true
+		}
+	}
+
+	if !foundRegion {
+		clusterDeploymentLog.Warnf("no AMI defined for cluster version %s/%s in region %s", clusterVersion.Namespace, clusterVersion.Name, clusterDeployment.Spec.Hardware.AWS.Region)
+		return fmt.Errorf("no AMI defined for cluster version %s/%s in region %s", clusterVersion.Namespace, clusterVersion.Name, clusterDeployment.Spec.Hardware.AWS.Region)
+	}
+	return nil
+}
+
+// setInvalidRegionStatus updates the cluster deployment status to indicate that the clusterVersion does not include an AMI for the region of
+// the cluster deployment
+func (c *Controller) setInvalidRegionStatus(original *clustop.ClusterDeployment, clusterVersion *clustop.ClusterVersion) error {
+	clusterDeployment := original.DeepCopy()
+	controller.SetClusterCondition(&clusterDeployment.Status, clustop.ClusterVersionIncompatible,
+		corev1.ConditionTrue,
+		versionMissingRegion,
+		fmt.Sprintf("no AMI defined for cluster version %s/%s in region %v", clusterVersion.Namespace, clusterVersion.Name, clusterDeployment.Spec.Hardware.AWS.Region),
+		controller.UpdateConditionIfReasonOrMessageChange)
+	return controller.PatchClusterDeploymentStatus(c.clustopClient, original, clusterDeployment)
+}
+
+// resolveClusterVersion checks if the cluster version referenced by the cluster deployment exists and returns it.
+// If not found or the lookup fails, an error is returned.
+func (c *Controller) resolveClusterVersion(clusterDeployment *clustop.ClusterDeployment) (*clustop.ClusterVersion, error) {
+	// Namespace may have been left empty signalling to use the clusterDeployment's namespace to locate the version:
+	versionNS := clusterDeployment.Spec.ClusterVersionRef.Namespace
+	if versionNS == "" {
+		versionNS = clusterDeployment.Namespace
+	}
+
+	cv, err := c.clusterVersionsLister.ClusterVersions(versionNS).Get(
+		clusterDeployment.Spec.ClusterVersionRef.Name)
+	return cv, err
+}
+
+func getMachineSetKey(machineSet *capi.MachineSet) string {
+	return fmt.Sprintf("%s/%s", machineSet.Namespace, machineSet.Name)
+}
+
+func getClusterKey(cluster *capi.Cluster) string {
+	return fmt.Sprintf("%s/%s", cluster.Namespace, cluster.Name)
+}
+
+func masterMachineSetName(clusterDeploymentName string) string {
+	return fmt.Sprintf("%s-master", clusterDeploymentName)
```

Comment:
We should use the name from the `MachineSetConfig` instead of using "master".

Reply:
@staebler I assume you mean the short name in the ClusterMachineSet. Afaik that name can change. Having a fixed name for the life of the cluster makes it easier to sync the machineset.

Reply:
@csrwng We should either (1) require the short name in the ClusterMachineSet to be "master", (2) not allow changes to the short name for the master, or (3) not allow the user to supply a name for the master at all and reserve the "master" name so that it cannot be used for compute machinesets. Of those options, I like (3) the best.